app/services/course/material/download_service.rb

```diff
@@ -0,0 +1,64 @@
+class Course::Material::DownloadService
+  require 'zip'
+  require 'uri'
+  require 'open-uri'
+
+  class << self
+    # Download the contents and zip them.
+    #
+    # @param [Array<Course::Material|Course::Material::Folder>] contents The contents to be
+    #   downloaded.
+    # @return [String] The url to the zip file.
+    def download_and_zip(contents)
+      tmp_dir = Dir.mktmpdir('coursemology-download-')
+
+      download(contents, tmp_dir)
+      zip_dir(tmp_dir)
+    end
+
+    private
+
+    # Download contents and store them in the given parent directory.
+    def download(contents, parent_dir)
+      contents = [contents].flatten
+      contents.each do |content|
+        if content.is_a?(Course::Material)
+          download_material(content, parent_dir)
+        else
+          sub_dir = File.join(parent_dir, sanitize_filename(content.name))
+          Dir.mkdir(sub_dir)
+          download(content.children + content.materials.includes(:attachments), sub_dir)
+        end
+      end
+    end
+
+    # Download the material and store it in the given parent directory.
+    def download_material(material, parent_dir)
+      local_file = File.join(parent_dir, sanitize_filename(material.name))
+      url = material.attachment.file_url
+      # Use path of file if url doesn't have host
+      source_file = URI(url).host ? url : material.attachment.file_path
+      IO.copy_stream(open(source_file), local_file)
+    end
+
+    def sanitize_filename(filename)
+      filename.gsub(/[:\/\\]/, ' ')
+    end
+
+    # Zip the directory and write to the file.
+    # @param [String] input_dir The path of the input directory.
+    #
+    # @return [String] The path of the zip file.
+    def zip_dir(input_dir)
+      output_filename = Dir::Tmpname.make_tmpname(['coursemology-download-', '.zip'], nil)
```

Comment:
I'm surprised I used two different ways to generate a temporary name... can we unify them somehow?